Std/Classes/DecidableLE.lean

```diff
@@ -0,0 +1,15 @@
+/--
+Asserts that the "less-than-or-equal" relation on `α` is decidable, that is, `a ≤ b` is decidable
+for all `a b : α`. See `Decidable`.
+
+It is generally not necessary to create instances of this class directly, because Lean will create
+one whenever a `DecidableRel` instance exists for the `LE` instance in question.
+-/
+class DecidableLE (α : Type u) extends LE α where
+  /-- Decide whether `x` is less than or equal to `y`. -/
+  decLE : (x y : α) → Decidable (LE.le x y)
+
+instance [DecidableLE α] {x y : α} : Decidable (x ≤ y) := DecidableLE.decLE _ _
+
+instance [inst : LE α] [DecidableRel (@LE.le α inst)] : DecidableLE α where
+  decLE := inferInstance
```

Comment:
I hope this will not make type class inference run into circles, or be significantly slower.

Reply:
I would like sign-off from mathlib maintainers for this

Reply:
Would an `abbrev` work better?

Reply:
Does
```lean
abbrev DecidableLE (α) {inst : LE α} := DecidableRel inst.le
```
work at all?

If you're proposing bundling `LE` and `DecidableRel` then I would favor something more basic like
```lean
class BLE (α) where ble : α → α → Bool
```
I have a vague feeling this has been debated abundantly elsewhere...

Reply:
Ah! This is tied to the long awaited `Decidable` overhaul. Since `Decidable` is not a structure (it's a rare `class inductive`) it can't easily be mixed in with other classes.

Reply:
RE `ble` - I think the important thing here is that people who are new to Lean can do basic operations like comparisons without understanding _all_ the subtleties of implicit arguments and dependent types that are implied by `[inst : LE α] [DecidableRel (@LE.le _ inst)]`.

Would `ble`/`blt` solve this? I would think that it would imply using some other version of `<`, just as `==` is the Boolean notion of equality. But I don't think users will have an easy time guessing non-standard comparison syntax.

Reply:
I think the core issue is that we can't use the `class abbrev` mechanism with `Decidable` since `Decidable` is a class inductive and not a class structure. This could be fixed by [lean4#2038](https://github.com/leanprover/lean4/pull/2038).

The circular dependency is probably going to slow down class inference for `LE`. Here are three less risky options:

1. A convenience abbreviation:
```lean
abbrev DecidableLE (α) [inst : LE α] := DecidableRel inst.le

def order [LE α] [DecidableLE α] : α × α → α × α
  | (x, y) => if x ≤ y then (x, y) else (y, x)
```

2. A non-circular version:
```lean
class DecidableLE (α) extends LE α where
  instDecidable (x y : α) : Decidable (x ≤ y)
attribute [instance] DecidableLE.instDecidable

def order [DecidableLE α] : α × α → α × α
  | (x, y) => if x ≤ y then (x, y) else (y, x)
```

3. A boolean version with no new notation:
```lean
class BLE (α) where ble : α → α → Bool
export BLE (ble)
instance [BLE α] : LE α where le x y := ble x y 
instance [BLE α] (x y : α) : Decidable (x ≤ y) := inferInstanceAs (Decidable (_ = true))

def order [BLE α] : α × α → α × α
  | (x, y) => if x ≤ y then (x, y) else (y, x)
```

Reply:
While we can't use `class abbrev` here, I think the result would be the same as this.
```lean
class DecidableLE (α : Type _) extends LE α where
  [instDecidable (x y : α) : Decidable (x ≤ y)]
attribute [instance] DecidableLE.mk DecidableLE.instDecidable
```

Reply:
Number 1 here imposes less, but still oddly large, syntactic overhead on people writing what should be basic programs. I can at least imagine how I'd write a good error message for that one!

Number 2 seems good - should I just update this PR to it? It'd be a bit annoying to write the instances, but not the end of the world. We're speculating about performance here, so it would be nice to compare e.g. the Mathlib build times on top of each option to find out rather than making educated guesses. Do we have a better way to benchmark std4's type classes?

The `BLE` one seems potentially problematic in the same ways that the split between `DecidableEq` and `BEq` are today. We could easily end up with two separate LE predicates for a type, with different behavior obtained only by replacing `BLE` with the pair of `LE` and `DecidableRel` in the signature. This strikes me as a footgun.

What's the best way to run real experiments here in `std4`? I think this goal is important but I'm by no means wedded to a particular implementation.

Reply:
I like the idea of solving this at the language level rather than library level.

There are a lot of classes related to binary ordering operations already and this feels a bit like adding one more to solve a particular problem.  I expect some of the Mathlib ones will flow down to Std eventually for theorems about ordered data structures.

The syntax for expressing the decidability of a binary operation isn't great though, so I'm not opposed to `DecidableLT α` as an abbreviation for `DecidableRel (α := α) (· < ·)` (or with the extra `inferInstance`) if that works out as a language feature.

Reply:
I agree that `abbrev DecidableLT (α : Type u) := DecidableRel (α := α) (· < ·)` would be a welcome addition; after all, we already have something analogous for `DecidableEq`.